providers/sum/sum.go

```diff
@@ -0,0 +1,280 @@
+}
+
+// Option for setting optional Client values
+type Option func(*Exec)
+
+func WithSumPath(sumPath string) Option {
+	return func(c *Exec) {
+		c.SumPath = sumPath
+	}
+}
+
+func WithLogger(log logr.Logger) Option {
+	return func(c *Exec) {
+		c.Log = log
+	}
+}
+
+func New(host, user, pass string, opts ...Option) (*Exec, error) {
+	sum := &Exec{
+		Host:     host,
+		Username: user,
+		Password: pass,
+		Log:      logr.Discard(),
+	}
+
+	for _, opt := range opts {
+		opt(sum)
+	}
+
+	var err error
+
+	if sum.SumPath == "" {
+		sum.SumPath, err = exec.LookPath("sum")
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		if _, err = os.Stat(sum.SumPath); err != nil {
+			return nil, err
+		}
+	}
+
+	return sum, nil
+}
+
+// Open a connection to a BMC
+func (c *Exec) Open(ctx context.Context) (err error) {
+	return nil
+}
+
+// Close a connection to a BMC
+func (c *Exec) Close(ctx context.Context) (err error) {
+	return nil
+}
+
+func (c *Exec) Name() string {
+	return ProviderName
+}
+
+func (s *Exec) run(ctx context.Context, command string, additionalArgs ...string) (output string, err error) {
+	// TODO(splaspood) use a tmp file here (as sum supports) to read the password
+	sumArgs := []string{"-i", s.Host, "-u", s.Username, "-p", s.Password, "-c", command}
+	sumArgs = append(sumArgs, additionalArgs...)
+
+	s.Log.V(9).WithValues(
+		"sumArgs",
+		sumArgs,
+	).Info("Calling sum")
+
+	cmd := exec.CommandContext(ctx, s.SumPath, sumArgs...)
+	b_out, err := cmd.CombinedOutput()
+	if err != nil || ctx.Err() != nil {
+		if err != nil {
+			return
+		}
+
+		if ctx.Err() != nil {
+			return output, ctx.Err()
+		}
+	}
+
+	return string(b_out), err
+}
+
+func (s *Exec) GetCurrentBiosCfg(ctx context.Context) (output string, err error) {
+	return s.run(ctx, "GetCurrentBiosCfg")
+}
+
+func (s *Exec) LoadDefaultBiosCfg(ctx context.Context) (err error) {
+	_, err = s.run(ctx, "LoadDefaultBiosCfg")
+	return
+}
+
+func (s *Exec) ChangeBiosCfg(ctx context.Context, cfgFile string, reboot bool) (err error) {
+	args := []string{"--file", cfgFile}
+
+	if reboot {
+		args = append(args, "--reboot")
+	}
+
+	_, err = s.run(ctx, "ChangeBiosCfg", args...)
+
+	return
+}
+
+// GetBiosConfiguration return bios configuration
+func (s *Exec) GetBiosConfiguration(ctx context.Context) (biosConfig map[string]string, err error) {
+	biosText, err := s.GetCurrentBiosCfg(ctx)
+	if err != nil {
+		return
+	}
+
+	// We need to call vcm here to take the XML returned by SUM and convert it into a simple map
+	vcm, err := config.NewVendorConfigManager("xml", common.VendorSupermicro, map[string]string{})
+	if err != nil {
+		return
+	}
+
+	err = vcm.Unmarshal(biosText)
+	if err != nil {
+		return nil, err
+	}
+
+	biosConfig, err = vcm.StandardConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	return biosConfig, nil
+}
+
+// SetBiosConfiguration set bios configuration
+func (s *Exec) SetBiosConfiguration(ctx context.Context, biosConfig map[string]string) (err error) {
```

Comment:
could we have some test cases here that verify the expected command is being executed